### PR TITLE
compiler: Improve blocking heuristics

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

**Review by georgebisbas — APPROVED** [review]

**Review by mloubout — APPROVED** [review]

File: tests/test_dimension.py [hunk 1]
```diff
@@ -1370,7 +1370,7 @@ def test_topofusion_w_subdims_conddims(self):
         eqns = [Eq(f.forward, f + 1),
                 Eq(g.forward, g + 1),
                 Eq(fsave, f.dt2, implicit_dims=[ctime]),
-                Eq(h, f + g, subdomain=grid.interior),
+                Eq(h, f.dx + g, subdomain=grid.interior),
```
Comment (georgebisbas) [inline tests/test_dimension.py:1373]: Why? Does this break with the new additions?

Comment (FabioLuporini) [reply tests/test_dimension.py:1373]: no, of course it doesn't "break" anything (why would it), but it's necessary to generate loop blocking due to these improvements: https://github.com/devitocodes/devito/pull/2016/files#diff-9a7b7fd6b3b90ced519c980132e20df6dabab8f794895fe9237f60880c80f006R98-R105

the alternative would have been to change the asserts. Perhaps equivalent.

Comment (georgebisbas) [reply tests/test_dimension.py:1373]: By "break", I mean the test would not pass

File: devito/passes/clusters/blocking.py [hunk 2]
```diff
@@ -95,13 +95,14 @@ def _process_fatd(self, clusters, level, prefix=None):
...
 
     def _has_data_reuse(self, cluster):
-        # A necessary but not sufficient condition for the existance of data
-        # reuse in the Cluster is that there must be at least three Indexeds --
-        # the LHS, the RHS, and another Indexed shifted w.r.t. the RHS, e.g.
-        # `a(x), b(x), b(x+1)`. Obv not sufficient because, e.g., `a(x), b(x),
-        # c(x)` would have no data reuse across x-iterations
-        if len(cluster.scope.indexeds) >= 3:
-            return True
+        # A sufficient condition for the existance of data reuse in `cluster`
```
Comment (georgebisbas) [inline devito/passes/clusters/blocking.py:98]: existence

**Comment by FabioLuporini** [comment]
I have to make changes so I'm closing this in favor to a new PR

**Pull request closed without merging**